[Excel](data types) New APIs for linked entities (#5091)
OfficeDev/office-js-docs-pr: docs/excel/excel-data-types-linked-entity-cell-values.md

---
title: Create linked entity cell values
description: Create linked entity cell values to represent large data sets in Excel.
ms.topic: how-to
ms.date: 03/31/2025
ms.localizationpriority: medium
---

# Create linked entity cell values (preview)

> 

Author: added

> 

Author: actually wait, this is in production Monday so don't need it now.

- Excel uses a cache to allow different cells to reference the same linked entity cell value seamlessly. This also improves workbook performance.

This article expands on information described in the following articles. We recommend reading the following articles before learning how to build your own linked entity cell values.

- [Excel data types: Stocks and geography](https://support.microsoft.com/office/excel-data-types-stocks-and-geography-61a33056-9935-484f-8ac8-f1a89e210877)
- [Overview of data types in Excel add-ins](excel-data-types-overview.md)
- [Excel JavaScript API data types entity value card](excel-data-types-entity-card.md).

## Key concepts

Linked entity cell values provide the user with data linked from an external data source. The user can view them as an entity value card.

:::image type="content" source="../images/excel-geography-linked-data-type-seattle.png" alt-text="Screen shot of an entity value data card for the Seattle Geography linked data type in the worksheet.":::

Like regular entity values, linked entity cell values can be referenced in formulas.

> :::image type="content" source="../images/excel-geography-seattle-dot-syntax.png" alt-text="Screen shot of using dot notation in a formula using =D6. To display fields for Seattle Geography data type.":::

Author: fixed

### Register the data domains

Let’s look at the code to register new domains named **Products** and **Categories**. The data provider name is **Contoso**. When the add-in loads, it first registers the data domains with Excel.

Use the [Excel.LinkedEntityDataDomainCreateOptions](/javascript/api/excel/excel.linkedentitydatadomaincreateoptions) type to describe the options you want, including which function to use as the linked entity load service. Then add the domain to the [Workbook.linkedEntityDataDomains](/javascript/api/excel/excel.workbook#excel-excel-workbook-linkedentitydatadomains-member) collection. It's recommended to register domains when you [Initialize your Office Add-in](../develop/initialize-add-in.md).
The following code shows how to register the **Products** and **Categories** data domains.

```javascript
Office.onReady(async () => {
  await Excel.run(async (context) => {
    const productsDomain: Excel.LinkedEntityDataDomainCreateOptions = {
      dataProvider: "Contoso",
      id: "products",
      name: "Products",
      // Id of the custom function that will be called on demand by Excel to resolve/refresh linked entity

>       // cell values of this data domain.

Reviewer: ```suggestion
```

    await context.sync();
  });
});
```

## Insert a linked entity cell value

There are two ways to insert a linked entity cell value into a cell on a worksheet.

- Create a command button on the ribbon or a button in your task pane. When the user selects the button your code inserts a linked entity cell value.
- Create a custom function that returns a linked entity cell value.

The following code example shows how to insert a new linked entity cell value into the selected cell. This code can be called from a command button on the ribbon, or a button in the task pane. Notes about the following code:

- You must specify a `serviceId` of `268436224` for any linked entity cell values you return. This informs Excel that the linked entity cell value is associated with an Excel add-in.
- You must specify a `culture`. Excel will pass it to your linked entity load service function so that you can maintain the original culture when the workbook is opened in a different culture.
- The `text` property is displayed to the user in the cell while the linked entity data value is updated. This prevents the user seeing a blank cell while the update is completed.

```javascript

> export async function insertProduct() {

Author: nope, fixed

    return null;
  }

  const categoryLinkedEntity: Excel.LinkedEntityCellValue = {
    type: "LinkedEntity",
    text: category.categoryName,
    id: {
      entityId: category.categoryID,
      domainId: categoriesDomainId,
      serviceId: addinDomainServiceId,
      culture: defaultCulture
    },
    properties: {
      "Category ID": {
        type: "String",
        basicValue: category.categoryID,
        propertyMetadata: {
          // Exclude the category ID property from the card view and auto complete.
          excludeFrom: {
            cardView: true,
            autoComplete: true
          }
        }
      },
      "Category Name": {
        type: "String",
        basicValue: category.categoryName
      },
      Description: {
        type: "String",
        basicValue: category.description
      }
    },
    layouts: {
      compact: {
        icon: "Branch"
      }
    }
  };

  return categoryLinkedEntity;
}
```

## Data refresh options

When you register a data domain, the user can refresh it manually at any time (such as choosing **Refresh All** from the **Data** tab.). There are three refresh modes you can specify for your data domain.

> - `manual`: The data is refreshed only when the user chooses to refresh. This is the default mode. Manual refresh can always be performed by the user, even when the refresh mode is set to `onLoad` or `periodic`.

Author: okay, dashes it is.

  // Manual refresh mode is always supported, even if unspecified.
  supportedRefreshModes: [
    Excel.LinkedEntityDataDomainRefreshMode.periodic,
    Excel.LinkedEntityDataDomainRefreshMode.onLoad
  ]
};
```

You can also programmatically request a refresh on a linked entity data domain by using either of the following methods.

- `LinkedEntityDataDomain.refresh()` - Refreshes all `LinkedEntityCellValue` objects of the linked entity data domain.
- `LinkedEntityDataDomainCollection.refreshAll()` - Refreshes all `LinkedEntityCellValue` objects of all linked entity data domains in the collection.

The refresh methods request a refresh which occurs asynchronously. To determine the results of the refresh, listen for the `onRefreshCompleted` event. The following code sample shows an example of listening for the `onRefreshCompleted` event.

```javascript
 await Excel.run(async (context) => {
    const dataDomains = context.workbook.linkedEntityDataDomains;
    dataDomains.onRefreshCompleted.add(onLinkedEntityDomainRefreshed);

    await context.sync();
});

async function onLinkedEntityDomainRefreshed (eventArgs: Excel.LinkedEntityDataDomainRefreshCompletedEventArgs): Promise<any> {

>   console.log("Linked entity domain refreshed: " + eventArgs.id);

Author: changed to 4 space to match rest of docs.

  console.log("Refresh status: " + eventArgs.refreshed);
  console.log("Refresh error: " + eventArgs.errors);
  return null;
}
```

## Error handling in linked entity load service

When Excel calls your add-in to get data for a linked entity cell value, it's possible an error can occur. If Excel is unable to connect to your add-in at all (such as when it is not loaded) it will display `#CONNECT!` to the user.

> If your linked entity load service function encounters an error, it should throw an error. The error will cause Excel to show `#CONNECT!` to the user.

Author: You don't interact with a load service (your CF is the load service) but I updated the text here.

The following code shows how to handle an error in a linked entity load service function.

```javascript
async function productLinkedEntityService(request: any): Promise<any> {
    const notAvailableError = new CustomFunctions.Error(CustomFunctions.ErrorCode.notAvailable);
    try {
      // Create an return a new linked entity cell value.
        let linkedEntityResult = ...
      ...
        if (!linkedEntityResult) {
        // Throw an error to signify to Excel that resolution/refresh of the requested linkedEntityId failed.
        throw notAvailableError;
      }
      ...
    } catch (error) {
        console.error(error);
        throw notAvailableError;
    }
}
```

## Debugging the linked entity load service

> Most add-in functionality for linked entity data types can be debugged using the guidance in [Overview of debugging Office Add-ins](../testing/debug-add-ins-overview.md). However, the linked entity load service function can be implemented in a shared runtime, or a JavaScript-only runtime (also know as a custom function runtime.) If you choose to implement the function in a JavaScript-only runtime, use the debugging guidance in JavaScript-only runtime for [Custom functions debugging in a non-shared runtime](custom-functions-debugging.md).

Reviewer: ```suggestion
Most add-in functionality for linked entity data types can be debugged using the guidance in [Overview of debugging Office Add-ins](../testing/debug-add-ins-overview.md). However, the linked entity load service function can be implemented in a shared runtime or a JavaScript-only runtime (also know as a custom functions runtime.) If you choose to implement the function in a JavaScript-only runtime, use the [Custom functions debugging in a non-shared runtime](custom-functions-debugging.md) guidance.
```

> The linked entity load service function uses the custom function architecture, regardless of which runtime you use. However, there are significant differences from regular custom functions.

Reviewer: ```suggestion
The linked entity load service function uses the custom functions architecture, regardless of which runtime you use. However, there are significant differences from regular custom functions.
```

Differences:

> - They will not appear to end users for usage in formulas.

Reviewer: What is "they" here? Linked entities? Is it specifically "Linked entity load service functions"?

Author: reworked the intros to be clearer.

- They do not support the JSDoc tags `@streaming` or `@volatile`.

Similarities:

- They use [Custom functions naming and localization](custom-functions-naming.md).
- They use the same error handling approach.

## Behavior in Excel 2019 and earlier

> If someone opens a worksheet with linked entity cell values on an older version of Excel that doesn’t support linked entity cell values, Excel will show the cell values as errors. This is the designed behavior. This is also why you set the `basicType` to `Error` and the `basicValue` to `#VALUE!` every time you insert or update a linked entity cell value. This is the error that Excel will fall back to on older versions.

Reviewer: ```suggestion
If someone opens a worksheet with linked entity cell values on an older version of Excel that doesn’t support linked entity cell values, Excel shows the cell values as errors. This is the designed behavior. This is also why you set the `basicType` to `Error` and the `basicValue` to `#VALUE!` every time you insert or update a linked entity cell value. This is the error that Excel uses as a fallback on older versions.
```

Author: fixed

## Best practices

> - Register linked entity data domains in the initialization code `Office.OnReady`, so that the user will have immediate functionality such as refreshing the linked entity cell values.

Reviewer: ```suggestion
- Register linked entity data domains in the initialization code `Office.OnReady` so that the user has immediate functionality and refreshed linked entity cell values.
```

Author: fixed

> - After publishing your add-in, don’t change the linked entity data domain ids. Consistent ids across logically the same objects will help with performance.

Reviewer: ```suggestion
- After publishing your add-in, don’t change the linked entity data domain IDs. Consistent IDs across the same logical objects helps with performance.
```

Author: fixed

> - Always provide the `text` property when creating a new linked entity cell value. This value is displayed while Excel calls your data provider function to get the remaining property values. Otherwise the user will see a blank cell until the data is retrieved.

Reviewer: ```suggestion
- Always provide the `text` property when creating a new linked entity cell value. This value is displayed while Excel calls your data provider function to get the remaining property values. Otherwise the user sees a blank cell until the data is retrieved.
```

Author: fixed

> don't use exclamation marks in product names.

Reviewer: Should this exclamation point line be in the preceding list? Or is it a copy paste error

Author: fixed (should be first)

Linked entity cell values are linked to an external data source. They provide the following advantages over regular entity values.

- Linked entity cell values can nest, and nested linked entity cell values are not retrieved until referenced; either by the user, or by the worksheet. This helps reduce file size and improve workbook performance.
- Excel uses a cache to allow different cells to reference the same linked entity cell value seamlessly. This also improves workbook performance.

This article expands on information described in the following articles. We recommend reading the following articles before learning how to build your own linked entity cell values.

> - [Excel data types: Stocks and geography](https://support.microsoft.com/office/excel-data-types-stocks-and-geography-61a33056-9935-484f-8ac8-f1a89e210877)

Reviewer 2: If you only use the GUID, the link still works if the SMC article gets renamed.
```suggestion
- [Excel data types: Stocks and geography](https://support.microsoft.com/office/61a33056-9935-484f-8ac8-f1a89e210877)
```
